test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetClientCertCommandTests.cs

```diff
@@ -0,0 +1,1033 @@
+                };
+            }
+
+            public TestInfo()
+            {
+                NuGetExePath = Util.GetNuGetExePath();
+                WorkingPath = TestDirectory.Create();
+                ConfigFile = Path.Combine(WorkingPath, "NuGet.config");
+                PackageSourceName = "Foo";
+                InvalidPackageSourceName = "Bar";
+                CertificateFileName = "contoso.pfx";
+                CertificateAbsoluteFilePath = Path.Combine(WorkingPath, CertificateFileName);
+                CertificateRelativeFilePath = ".\\" + CertificateFileName;
+                CertificatePassword = "password";
+
+                CertificateStoreLocation = StoreLocation.CurrentUser;
+                CertificateStoreName = StoreName.My;
+                CertificateFindBy = X509FindType.FindByIssuerName;
+                CertificateFindValue = "Patti Fuller";
+
+                Util.CreateFile(Path.GetDirectoryName(ConfigFile),
+                                Path.GetFileName(ConfigFile),
+                                $@"
+<configuration>
+    <packageSources>
+        <add key=""{PackageSourceName}"" value=""https://contoso.com/v3/index.json"" />
+    </packageSources>
+</configuration>
+");
+            }
+
+            public string CertificateAbsoluteFilePath { get; }
+            public string CertificateFileName { get; }
+            public X509FindType CertificateFindBy { get; }
+            public object CertificateFindValue { get; }
+            public string CertificatePassword { get; }
+            public string CertificateRelativeFilePath { get; }
+            public StoreLocation CertificateStoreLocation { get; }
+            public StoreName CertificateStoreName { get; }
+
+            public string ConfigFile { get; }
+            public string InvalidPackageSourceName { get; }
+            public string NuGetExePath { get; }
+
+            public string PackageSourceName { get; }
+            public TestDirectory WorkingPath { get; }
+
+            public void Dispose()
+            {
+                WorkingPath.Dispose();
+                RemoveCertificateFromStorage();
+            }
+
+            public void SetupCertificateFile()
+            {
+                var certificateData = CreateCertificate();
+                File.WriteAllBytes(CertificateAbsoluteFilePath, certificateData);
+            }
+
+            public void SetupCertificateInStorage()
+            {
+                using (var store = new X509Store(CertificateStoreName, CertificateStoreLocation))
+                {
+                    store.Open(OpenFlags.ReadWrite);
+                    var password = new SecureString();
+                    foreach (var symbol in CertificatePassword)
+                    {
+                        password.AppendChar(symbol);
+                    }
+
+                    store.Add(new X509Certificate2(CreateCertificate(), password));
+                }
+            }
```

Comment:
I'm not sure how I feel about tests that modify machine state, even if it's cleaned up at the end of the test. I'd be ok with only testing store certs manually.

@heng-liu has been working on getting package signature validations working cross-platform. I remember an email from the dotnet crypto experts, where they said something like certs can take seconds from when they're added in the store class, to when they're available for querying. Unforunately I can't find the email at the moment. Therefore, if the test tries to search the store for the cert too quickly after it was added, it could fail, causing false negatives.

Additionally, this helper class is used for all the tests, but only a few of them use the cert store, and I really don't feel comfortable spamming the cert store with new certs and deleting them when we don't even use it.

@heng-liu @zkat @nkolev92 What do you think? Should we have tests for adding a store cert reference into our config file?

At least the package signature tests actually use the cert and doing crypto things with it. These tests just check it exists.

Reply:
On Linux, there is a latency between you add certificate to store and the certificate is trusted. The related issue is https://github.com/dotnet/runtime/issues/32608

Reply:
Is there something coming soon that we can use to do this better? ["or use the .NET 5 custom trust store feature"](https://github.com/dotnet/runtime/issues/32608#issuecomment-589763492)

Reply:
Looking at the issue that Heng linked (thanks Heng!), the delay issue doesn't apply because the cert is added into one X509Store instance, and then a different instance is searching for it, and the issue says that `new X509Store` will always use live data.

Affecting machine state still feels wrong to me, but it's not as big an issue as I first thought.

At a minimum, I do ask that the utility class only creates the cert in the store when it's actually needed, not every time. Consider a constructor argument about whether or not to create the cert in the store, and only set it to true for tests that actually search the store.